### PR TITLE
Graceful worker shutdown

### PR DESCRIPTION
**Pull request by mdellweg**
fixes #8930
https://pulp.plan.io/issues/8930

### CODE REVIEW

File: pulpcore/tasking/pulpcore_worker.py [hunk 1]
```diff
@@ -162,14 +174,25 @@ def supervise_task(self, task):
                     break
                 if not task_process.is_alive():
                     break
+            if self.shutdown_requested:
+                grace -= 1
+                if grace > 0:
+                    _logger.info(f"Gracefully wait for task {task.pk} to finish.")
```
Comment (dralley) [inline pulpcore/tasking/pulpcore_worker.py:180]: Would these messages be better rewritten as "Shutdown requested, waiting for task {task.pk} to finish." and "Aborting current task {task.pk} due to worker shutdown."?

File: pulpcore/tasking/pulpcore_worker.py [hunk 2]
```diff
@@ -162,14 +174,25 @@ def supervise_task(self, task):
                     break
                 if not task_process.is_alive():
                     break
+            if self.shutdown_requested:
+                grace -= 1
+                if grace > 0:
+                    _logger.info(f"Shutdown requested, waiting for task {task.pk} to finish.")
```
Comment (dralley) [inline pulpcore/tasking/pulpcore_worker.py:180]: We should probably say "Worker shutdown requested" to be as specific as possible and consistent with what is up above.

**Review by dralley — APPROVED** [review]

**Comment by bmbouter** [comment]
I'm a little concerned about the workers not shutting down quick enough if there is no work being handled. When I shut down workers with systemd no for example I see:

```
Jun 28 16:59:46 pulp3-source-fedora34.localhost.example.com pulpcore-worker[347230]: pulp [None]: pulpcore.tasking.pulpcore_worker:INFO: Worker 347230@pulp3-source-fedora34.localhost.example.com was requested to shut down.
Jun 28 16:59:51 pulp3-source-fedora34.localhost.example.com pulpcore-worker[347230]: pulp [None]: pulpcore.tasking.pulpcore_worker:INFO: Worker 347230@pulp3-source-fedora34.localhost.example.com was shut down.
```

Without this the workers shutoff immediately. What can be done to improve the speed of graceful shutdown when a worker is idle?

**Comment by bmbouter** [comment]
Without this PR when I tested shutdown of all kinds, the content app still shows workers online until the 30 second timeout occurs.

With this PR I see:
* shutdown via `systemd stop` takes a little bit of time to occur, but once it occurs the worker record is removed
* interactive shutdown via Ctrl-C does not work as expected (to me). Here's what I see:

1. Start the API and content apps so you can run `pulp status | jq '. | {online_workers}'`
2. Start `pulpcore-worker` interactively
3. Verify it's shown with `pulp status | jq '. | {online_workers}'`
4. Ctrl-C and see the worker shutdown immediately
5. Observe the worker record is still there for 30 seconds in `pulp status | jq '. | {online_workers}'`.

**Comment by bmbouter** [comment]
Shutting down an idle worker interactively and via-systemd is working nicely with this PR. When I monitor the status API, whenever those processes exit the status API reflects it immediately.

**Comment by bmbouter** [comment]
I was testing interactive shutdown for a worker working on a task, and I get this traceback each time.

```
(pulp) [vagrant@pulp3-source-fedora34 ~]$ pulpcore-worker
pulp [None]: pulpcore.tasking.entrypoint:INFO: Starting distributed type worker
pulp [None]: pulpcore.tasking.worker_watcher:INFO: New worker '352288@pulp3-source-fedora34.localhost.example.com' discovered
pulp [c67e8b8e8b094f4996c004b4db42be85]: pulpcore.tasking.pulpcore_worker:INFO: Starting task 212ddd50-114f-471d-a54f-076d277c77c4
^Cpulp [c67e8b8e8b094f4996c004b4db42be85]: pulpcore.tasking.pulpcore_worker:INFO: Worker 352288@pulp3-source-fedora34.localhost.example.com was requested to shut down.
pulp [None]: pulpcore.tasking.pulpcore_worker:INFO: Worker 352288@pulp3-source-fedora34.localhost.example.com was requested to shut down.
pulp [None]: pulpcore.tasking.pulpcore_worker:INFO: Worker shutdown requested, waiting for task 212ddd50-114f-471d-a54f-076d277c77c4 to finish.
pulp [None]: pulpcore.tasking.pulpcore_worker:INFO: Worker shutdown requested, waiting for task 212ddd50-114f-471d-a54f-076d277c77c4 to finish.
pulp [c67e8b8e8b094f4996c004b4db42be85]: pulpcore.tasking.pulpcore_worker:INFO: Task completed 212ddd50-114f-471d-a54f-076d277c77c4
Traceback (most recent call last):
  File "/usr/local/lib/pulp/bin/pulpcore-worker", line 33, in <module>
    sys.exit(load_entry_point('pulpcore', 'console_scripts', 'pulpcore-worker')())
  File "/usr/local/lib/pulp/lib64/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/pulp/lib64/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/pulp/lib64/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/pulp/lib64/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/vagrant/devel/pulpcore/pulpcore/tasking/entrypoint.py", line 41, in worker
    NewPulpWorker().run_forever()
  File "/home/vagrant/devel/pulpcore/pulpcore/tasking/pulpcore_worker.py", line 210, in run_forever
    self.shutdown()
  File "/home/vagrant/devel/pulpcore/pulpcore/tasking/pulpcore_worker.py", line 59, in shutdown
    self.worker.delete()
  File "/usr/local/lib/pulp/lib64/python3.9/site-packages/django_lifecycle/mixins.py", line 146, in delete
    value = super().delete(*args, **kwargs)
  File "/usr/local/lib/pulp/lib64/python3.9/site-packages/django/db/models/base.py", line 915, in delete
    assert self.pk is not None, (
AssertionError: Worker object can't be deleted because its pulp_id attribute is set to None.
```

To reproduce:
1. Start the content app and API app
2. Start a pulp-worker interactively
3. Modify pulp_file sync to have a 20 second sleep statement
4. Sync pulp_file and watch the task start and get stuck on the sleep
5. Ctrl-C the interactive worker
6. Observe that traceback

**Comment by mdellweg** [comment]
> I'm a little concerned about the workers not shutting down quick enough if there is no work being handled. When I shut down workers with systemd no for example I see:
> 
> ```
> Jun 28 16:59:46 pulp3-source-fedora34.localhost.example.com pulpcore-worker[347230]: pulp [None]: pulpcore.tasking.pulpcore_worker:INFO: Worker 347230@pulp3-source-fedora34.localhost.example.com was requested to shut down.
> Jun 28 16:59:51 pulp3-source-fedora34.localhost.example.com pulpcore-worker[347230]: pulp [None]: pulpcore.tasking.pulpcore_worker:INFO: Worker 347230@pulp3-source-fedora34.localhost.example.com was shut down.
> ```
> 
> Without this the workers shutoff immediately. What can be done to improve the speed of graceful shutdown when a worker is idle?

I do not know how. Python's `select` does not trigger on signals (like the underlying syscall). I suspect this to be solved, once we rewrite the outer worker code with python async. Until then a heartbeat is the timeframe on which we capture a signal.

**Comment by mdellweg** [comment]
> I was testing interactive shutdown for a worker working on a task, and I get this traceback each time.
> 
> ```
> (pulp) [vagrant@pulp3-source-fedora34 ~]$ pulpcore-worker
> pulp [None]: pulpcore.tasking.entrypoint:INFO: Starting distributed type worker
> pulp [None]: pulpcore.tasking.worker_watcher:INFO: New worker '352288@pulp3-source-fedora34.localhost.example.com' discovered
> pulp [c67e8b8e8b094f4996c004b4db42be85]: pulpcore.tasking.pulpcore_worker:INFO: Starting task 212ddd50-114f-471d-a54f-076d277c77c4
> ^Cpulp [c67e8b8e8b094f4996c004b4db42be85]: pulpcore.tasking.pulpcore_worker:INFO: Worker 352288@pulp3-source-fedora34.localhost.example.com was requested to shut down.
> pulp [None]: pulpcore.tasking.pulpcore_worker:INFO: Worker 352288@pulp3-source-fedora34.localhost.example.com was requested to shut down.
> pulp [None]: pulpcore.tasking.pulpcore_worker:INFO: Worker shutdown requested, waiting for task 212ddd50-114f-471d-a54f-076d277c77c4 to finish.
> pulp [None]: pulpcore.tasking.pulpcore_worker:INFO: Worker shutdown requested, waiting for task 212ddd50-114f-471d-a54f-076d277c77c4 to finish.
> pulp [c67e8b8e8b094f4996c004b4db42be85]: pulpcore.tasking.pulpcore_worker:INFO: Task completed 212ddd50-114f-471d-a54f-076d277c77c4
> Traceback (most recent call last):
>   File "/usr/local/lib/pulp/bin/pulpcore-worker", line 33, in <module>
>     sys.exit(load_entry_point('pulpcore', 'console_scripts', 'pulpcore-worker')())
>   File "/usr/local/lib/pulp/lib64/python3.9/site-packages/click/core.py", line 829, in __call__
>     return self.main(*args, **kwargs)
>   File "/usr/local/lib/pulp/lib64/python3.9/site-packages/click/core.py", line 782, in main
>     rv = self.invoke(ctx)
>   File "/usr/local/lib/pulp/lib64/python3.9/site-packages/click/core.py", line 1066, in invoke
>     return ctx.invoke(self.callback, **ctx.params)
>   File "/usr/local/lib/pulp/lib64/python3.9/site-packages/click/core.py", line 610, in invoke
>     return callback(*args, **kwargs)
>   File "/home/vagrant/devel/pulpcore/pulpcore/tasking/entrypoint.py", line 41, in worker
>     NewPulpWorker().run_forever()
>   File "/home/vagrant/devel/pulpcore/pulpcore/tasking/pulpcore_worker.py", line 210, in run_forever
>     self.shutdown()
>   File "/home/vagrant/devel/pulpcore/pulpcore/tasking/pulpcore_worker.py", line 59, in shutdown
>     self.worker.delete()
>   File "/usr/local/lib/pulp/lib64/python3.9/site-packages/django_lifecycle/mixins.py", line 146, in delete
>     value = super().delete(*args, **kwargs)
>   File "/usr/local/lib/pulp/lib64/python3.9/site-packages/django/db/models/base.py", line 915, in delete
>     assert self.pk is not None, (
> AssertionError: Worker object can't be deleted because its pulp_id attribute is set to None.
> ```
> 
> To reproduce:
> 
>     1. Start the content app and API app
> 
>     2. Start a pulp-worker interactively
> 
>     3. Modify pulp_file sync to have a 20 second sleep statement
> 
>     4. Sync pulp_file and watch the task start and get stuck on the sleep
> 
>     5. Ctrl-C the interactive worker
> 
>     6. Observe that traceback

I was unable to reproduce this behaviour.

**Comment by mdellweg** [comment]
@bmbouter as per your request, i added proper signal handling and task cancel timing. It really feels like reimplementing async python now. As i think this last minute change is kind of dangerous, i made it a separate commit. Please have a look. I can squash it later if requested.

**Comment by bmbouter** [comment]
> @bmbouter as per your request, i added proper signal handling and task cancel timing. It really feels like reimplementing async python now. As i think this last minute change is kind of dangerous, i made it a separate commit. Please have a look. I can squash it later if requested.

Thanks for showing that. I was thinking there was an easier way, but I didn't actually think about how. I also think this last minute change is kind of dangerous, so I'd prefer to go with what you had before if that's alright. I think we should prioritize an issue to convert this to async post 3.14. GA. Also is there an issue filed for the convert to async work, and if not, could you file one?

What do you think about this plan and request?

**Comment by mdellweg** [comment]
> > @bmbouter as per your request, i added proper signal handling and task cancel timing. It really feels like reimplementing async python now. As i think this last minute change is kind of dangerous, i made it a separate commit. Please have a look. I can squash it later if requested.
> 
> Thanks for showing that. I was thinking there was an easier way, but I didn't actually think about how. I also think this last minute change is kind of dangerous, so I'd prefer to go with what you had before if that's alright. I think we should prioritize an issue to convert this to async post 3.14. GA. Also is there an issue filed for the convert to async work, and if not, could you file one?
> 
> What do you think about this plan and request?

Now i think, we need to take this, as it turned out the signals were always bubbled to the child processes, so the only way (i found) to get both graceful shutdown as well as proper task cancelation working is by signalling the child process with SIGUSR1.

**Review by dralley — APPROVED** [review]

**Comment by bmbouter** [comment]
This is tests really nicely for me:
1. interactive Ctrl-C shuts down the worker immediately
2. systemd shutdown shuts down the worker immediately
3. interactive or systemd shutdown remove the worker from that status API immediately (monitored with `watch -n 0.5 "pulp status | jq '. | {online_workers}'"`).
4. shutting down a worker busy with a task, does wait for a while and the forcefully abort (which is good)
5. a double Crtl-C sent to a busy worker, does cause the worker process to exit immediately and not way

I did notice two small issues with this PR, but I want to treat them as separate things so that we can accept the good things this PR brings. I'm going to file those separately.

**Comment by bmbouter** [comment]
Here are the issues I filed:
https://pulp.plan.io/issues/8986
https://pulp.plan.io/issues/8987
https://pulp.plan.io/issues/8988

**Pull request merged**